src/main/java/frc/robot/subsystems/DriveSubsystem.java

```diff
@@ -0,0 +1,124 @@
+        pigeon2 = new Pigeon2(RobotMap.PIGEON_ID);
+        leftBackMotor.configFactoryDefault();
+        rightBackMotor.configFactoryDefault();
+        leftFrontMotor.configFactoryDefault();
+        rightFrontMotor.configFactoryDefault();
+        pigeon2.getConfigurator().apply(new Pigeon2Configuration());
+
+        leftFrontMotor.setInverted(true);
+        leftBackMotor.setInverted(true);
+        leftFrontMotor.setSensorPhase(true);
+
+        this.field2d = new Field2d();
+
+        differentialDriveOdometry = new DifferentialDriveOdometry(
+                new Rotation2d(getAngleDegrees()),
+                getLeftDistancePassedMeters(),
+                getRightDistancePassedMeters()
+        );
+        field2d.setRobotPose(differentialDriveOdometry.getPoseMeters().getX(), differentialDriveOdometry.getPoseMeters().getY(), differentialDriveOdometry.getPoseMeters().getRotation());
+
+        initialize();
+
+
+    }
+
+    public Field2d getField2d() {
+        return field2d;
+    }
+
+    public double getLeftDistancePassedMeters() {
+        return leftFrontMotor.getSelectedSensorPosition() / RobotMap.TALON_ENCODER_PPR * RobotMap.DRIVE_WHEEL_RADIUS * 2 * Math.PI;
+    }
+
+    public double getRightDistancePassedMeters() {
+        return rightBackMotor.getSelectedSensorPosition() / RobotMap.TALON_ENCODER_PPR * RobotMap.DRIVE_WHEEL_RADIUS * 2 * Math.PI;
+    }
+
+    public void initialize() {
+        pigeon2.reset();
+        leftFrontMotor.setSelectedSensorPosition(0);
+        rightBackMotor.setSelectedSensorPosition(0);//inits
+    }
+
+    public double getAngleDegrees() {
+        return (360 - pigeon2.getAngle()) % 360;// the value returned will be from 0 - 360 depending on its location
+        // 0  - degree after initializing, 359 - one degree to the right, 1 - one degree to the left// 90 will be 90 degrees to the left
+    }
+
+    public void powerLeftMotors(double powerL) {
+        leftFrontMotor.set(powerL);
+        leftBackMotor.set(powerL);
+    }
+
+    public void powerRightMotors(double powerR) {
+        rightFrontMotor.set(powerR);
+        rightBackMotor.set(powerR);
+    }
+
+    public void stop() {
+        rightBackMotor.stopMotor();
+        rightFrontMotor.stopMotor();
+        leftBackMotor.stopMotor();
+        leftFrontMotor.stopMotor();
+    }
+
+    //public void setOdometryPose2d (Pose2d pose2d){
+    //   differentialDriveOdometry.update(pose2d.getRotation(), getRightDistancePassedMeters(), getLeftDistancePassedMeters());
+    //}
+
+    private void updateOdometry() {
+        differentialDriveOdometry.update(
+                pigeon2.getRotation2d(),
+                getLeftDistancePassedMeters(),
+                getRightDistancePassedMeters()
+        );
+    }
+
+    public void periodic() {
+        differentialDriveOdometry.update(new Rotation2d(getAngleDegrees()), getLeftDistancePassedMeters(), getRightDistancePassedMeters());
```

Comment:
you should check the odometery on the field, you will find that you have problems with orientation. It is related to how you work with degrees with the odometery.

That's all I'll say for now

Reply:
**tried to fix.**